Fixing #5054, public_key:pkix_sign for RSASSA-PSS signature

sigsergv opened the pull request:
fixes #5054

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/erlang/otp?pullRequest=5057) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/erlang/otp?pullRequest=5057) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/erlang/otp?pullRequest=5057) it.</sub>

sigsergv commented:
Proposed fix for issue #5054

IngelaAndin commented:
Thanks, could you base it on maint instead of master? As it is a bug fix it would be nice to have it in the OTP-24 track. It looks like it is not breaking any existing test cases,  which I think is expected. Although the existing test cases all use a salt length that is also compliant with the old implementation, so please consider adding a test case using a different salt length.

sigsergv commented:
BTW, looks like test `rsa_pss_sign_verify` in `public_key_SUITE.erl` does nothing, there is no actual check:

```
public_key:pkix_verify(Cert, {#'RSAPublicKey'{modulus=Mod, publicExponent=Exp}, Parms}).
```

It should be:

```
true = public_key:pkix_verify(Cert, {#'RSAPublicKey'{modulus=Mod, publicExponent=Exp}, Parms}).
```

And it fails then because this certificate is created incorrectly, it's an EC certificate, not rsa-pss. That is not directly related to this PR or the original bug, so I created a separate PR: https://github.com/erlang/otp/pull/5064

IngelaAndin commented:
Good catch!  This PR  does not fail our TLS tests that use correct RSA-PSS certs :) But of course, we want the public_key test to be correct too. It would still be nice if you could rebase both your PRs so that they are based on the maint branch and will make it for OTP-24.1 and not on the master branch that will become OTP-25 (much later). Of course, maint will be merged to master so that it is fixed in both tracks.

sigsergv commented:
I made #5064 based on main branch. And still working on #5057.

IngelaAndin commented:
Ok, thank you!

sigsergv commented:
Added additional check into public_key_SUITE:pkix_path_validation/1

Also found and fixed the same issue in public_key:pkix_verify

Commits are rebased on maint.

IngelaAndin commented on lib/public_key/test/public_key_SUITE.erl:
```diff
@@ -1355,6 +1373,6 @@ pss_params(sha256) ->
                                               parameters = #'HashAlgorithm'{algorithm = ?'id-sha256'}
                                              },
-       saltLength = 32,
+       saltLength = 20,
```
It is recommended that the MGF hash function be the same as the scheme hash algorithm/function and that the salt length is hLen, the length of the output of the hash function.

sigsergv commented:
Yup, forgot to restore it after playing with non-standard parameters.

Fixed.

IngelaAndin commented:
Added lates to the test cycle again

IngelaAndin commented:
Thank you for the PR.
